src/google_cloud_debugger/google_cloud_debugger_lib/dbg_class_field.cc

```diff
@@ -170,6 +179,100 @@ HRESULT DbgClassField::Evaluate(
   return S_OK;
 }
 
+HRESULT DbgClassField::ProcessConstField(ICorDebugModule *debug_module,
+                                         IMetaDataImport *metadata_import) {
+  CorElementType field_type = (CorElementType)default_value_type_flags_;
+  ULONG64 enum_numerical_value = 0;
+  std::unique_ptr<DbgObject> dbg_object;
+  HRESULT hr = obj_factory_->CreateDbgObjectFromLiteralConst(
+      field_type, default_value_, default_value_len_, &enum_numerical_value,
+      &dbg_object);
+
+  if (FAILED(hr)) {
+    return hr;
+  }
+
+  // If this is an enum, the signature length will be greater than 2
+  // and the second byte will be CorElementType::ELEMENT_TYPE_VALUETYPE.
+  CorElementType second_byte_type =
+      (CorElementType) * (signature_metadata_ + 1);
+  if (second_byte_type != CorElementType::ELEMENT_TYPE_VALUETYPE) {
+    member_value_ = std::move(dbg_object);
+    return S_OK;
+  }
+
+  return ProcessConstEnumField(debug_module, metadata_import, field_type,
+                               enum_numerical_value);
+}
+
+HRESULT DbgClassField::ProcessConstEnumField(ICorDebugModule *debug_module,
+                                             IMetaDataImport *metadata_import,
+                                             const CorElementType &enum_type,
+                                             ULONG64 enum_numerical_value) {
+  // Gets the class that implements this field.
+  std::string class_name;
+  mdToken base_token;
+  HRESULT hr = debug_helper_->GetTypeNameFromMdTypeDef(
+      parent_token_, metadata_import, &class_name, &base_token,
+      GetErrorStream());
+
+  // Retrieves the parent class of the class that implements this field.
+  CorTokenType token_type = (CorTokenType)(TypeFromToken(base_token));
+  std::string base_class_name;
+  mdToken base_class_parent_token;
+  if (token_type == CorTokenType::mdtTypeDef) {
+    hr = debug_helper_->GetTypeNameFromMdTypeDef(
+        base_token, metadata_import, &base_class_name, &base_class_parent_token,
+        GetErrorStream());
+  } else if (token_type == CorTokenType::mdtTypeRef) {
+    hr = debug_helper_->GetTypeNameFromMdTypeRef(
+        base_token, metadata_import, &base_class_name, GetErrorStream());
+  }
+
+  // Do not proceed if the class implementing this field is
+  // an enum (which means the parent is System.Enum).
+  // This can lead to an infinite loop.
+  if (base_class_name.compare(kEnumClassName) != 0) {
+    // We need to get the enum name. The name of the enum can be retrieved
+    // from the third byte of the metadata signature.
+    if (sig_metadata_length_ < 3) {
```

Comment:
I think it might be worth putting all of these manipulations and checks into a separate class with static functions.  I think it would make things easier to read/follow and to test.